app/models/stacks_media_token.rb

```diff
@@ -0,0 +1,98 @@
+# minted it), the recovered hash can be checked against the provided list of values
+# and the current time.  if the expected values match the decrypted values and the
+# token is younger than the max age, it's good, otherwise, it's not.
+#
+# thus, stacks can mint a token string and give it to a user, who can give it
+# to a service that stacks trusts, to allow that service to fulfil a request
+# to stream a resource.
+#
+# the order of operations might be something like the following example:
+# * a user makes a request to a stacks URL for a streaming resource.
+# * stacks creates a StacksMediaToken and serializes it to its encrypted string form.
+# * the encrypted string form of the token is included as part of a redirect to wowza.
+# * wowza receives the request for the resource stream, with the encrypted string.
+# * wowza calls back to stacks, and passes along: encrypted token string, IP from which
+#   it received the request, id, and file name.
+# * stacks uses StacksMediaToken#verify_encrypted_token_string to determine whether
+#   the encrypted string represents a still-valid token for the given resource.
+# * if it does, it returns 200 OK to wowza, if it doesn't then it returns 403 Forbidden.
+# * wowza serves the stream or not based on the response it gets from stacks.
+# * this checking in wowza can be implemented via a wowza request processing plugin.
+#
+# note that there's no check for whether a token's been used (there's no record of the token
+# minting in memory or DB, and so nothing to mark).
+class StacksMediaToken
+  attr_reader :id, :file_name, :user_ip_addr, :timestamp
+
+  def initialize(id, file_name, user_ip_addr)
+    @id = id
+    @file_name = file_name
+    @user_ip_addr = user_ip_addr
+    @timestamp = Time.zone.now
+  end
+
+  def to_hash
+    {
+      id: id,
+      file_name: file_name,
+      user_ip_addr: user_ip_addr,
+      timestamp: timestamp
+    }
+  end
+
+  # though the object is instantiated by the constructor, the string returned by this
+  # method is what "mints" the token for the purposes of checks from other services
+  def to_encrypted_string
+    self.class.encryptor.encrypt_and_sign to_hash
+  end
+
+  # the token is valid if
+  #  * it has the values specified by the `expected_` params
+  #  * it hasn't yet expired
+  # NOTE: since this method does an expiry check at the time it's run, its result can
+  # become stale, and should not be cached.
+  def token_still_valid_for?(expected_id, expected_file_name, expected_user_ip_addr)
+    # max_token_age returns a duration.  calling `.ago` on it returns a date which we can check against for expiry.
+    id == expected_id && file_name == expected_file_name && user_ip_addr == expected_user_ip_addr &&
+      (timestamp >= self.class.max_token_age.ago)
+  end
+
+  def self.max_token_age
+    Settings.stream.max_token_age.to_i.seconds
+  end
+
+  def self.create_from_hash(smt_hash)
+    new smt_hash[:id], smt_hash[:file_name], smt_hash[:user_ip_addr]
```

Comment:
i like that `new` calls out the expected parameters, since we decided to forgo allowing the token payload to be an arbitrary hash.  but i can't really think of any benefits other than that.